src/loaders/compute_tools/microtrait/microtrait_util.py

```diff
@@ -0,0 +1,59 @@
+from collections import defaultdict
+from pathlib import Path
+
+import pandas as pd
+
+from src.loaders.common.loader_common_names import SYS_TRAIT_ID
+
+# column names from the rule2trait file that contain the trait names
+_RULE2TRAIT_TRAIT_NAME_COLS = ['microtrait_trait-name1', 'microtrait_trait-name2', 'microtrait_trait-name3']
+_RULE2TRAIT_RULE_NAME_COL = 'microtrait_rule-name'  # column name from the rule2trait file that contains the rule name
+
+# `microtrait_ruleunwrapped.txt` file unfortunately has no header, so we need to specify the column names
+_UNWRAPPED_RULE_COL = 'unwrapped_rule'
+_RULEUNWRAPPED_RULE_NAME_COL = 'rule_name'
+_RULEUNWRAPPED_COLS = [
+    _RULEUNWRAPPED_RULE_NAME_COL,
+    'rule_boolean',
+    'rule_display_name',
+    _UNWRAPPED_RULE_COL]
+
+
+def create_trait_unwrapped_rules(
+        rule2trait_file: Path,
+        ruleunwrapped_file: Path,
+        trait_unwrapped_rules_file: Path = None,
+) -> dict[str, set[str]]:
+    """
+    Generate a mapping of traits to unwrapped rule expressions.
+    """
+    trait_rule_unwrapped_mapping = defaultdict(set)
+
+    # read the rule2trait file and populate rule names to each trait
+    rule2trait_df = pd.read_csv(rule2trait_file, sep='\t')
+    for index, row in rule2trait_df.iterrows():
+        rule_name = row[_RULE2TRAIT_RULE_NAME_COL]
+        for trait_name_col in _RULE2TRAIT_TRAIT_NAME_COLS:
+            trait_name = row[trait_name_col]
+            trait_rule_unwrapped_mapping[trait_name].add(rule_name) if pd.notna(trait_name) else None
+
+    # read the ruleunwrapped file and populate unwrapped rule expressions to each trait
+    ruleunwrapped_df = pd.read_csv(ruleunwrapped_file, sep='\t', names=_RULEUNWRAPPED_COLS)
+    for trait_name, rule_names in trait_rule_unwrapped_mapping.items():
+        for rule_name in rule_names.copy():
+            unwrapped_rule = ruleunwrapped_df.loc[
+                ruleunwrapped_df[_RULEUNWRAPPED_RULE_NAME_COL] == rule_name, _UNWRAPPED_RULE_COL].values
+            if len(unwrapped_rule) > 0:
+                # remove the rule name from the set and add the unwrapped rule expression
+                trait_rule_unwrapped_mapping[trait_name].remove(rule_name)
+                trait_rule_unwrapped_mapping[trait_name].add(unwrapped_rule[0])
+
+    if trait_unwrapped_rules_file is not None:
+        with open(trait_unwrapped_rules_file, 'w') as file:
+            file.write(f"{SYS_TRAIT_ID}\t{_UNWRAPPED_RULE_COL}\n")
+            for trait, rule_set in trait_rule_unwrapped_mapping.items():
+                # Join the set elements into a single string
+                unwrapped_rules = ';'.join(rule_set)
+                file.write(f"{trait}\t{unwrapped_rules}\n")
```

Comment:
Can you provide an example of what the contents of the output file look like?

Reply:
just two columns with trait_id and associated unwrapped rule. 
`trait_id	unwrapped_rule
Resource Acquisition:Substrate degradation:simple compound degradation:carbohydrate degradation	((('galM')) & (('galK')) & (('galT')) & (('galE' | 'capD')));(('manA'));(('treS1'));((('fucI')) & (('fucK')) & (('fucA')));(('glk')) & (('gpd')) & (('pgl_1' | 'pgl_2')) & (('edd')) & (('eda' | 'dgaF' | 'kdpgA'));(('malQ' | 'jgt'));(('glk')) & (('pgi' | 'pgi_arch' | 'pgi-tal' | 'pgi-pmi')) & (('pfk' | 'pfkA' | 'pfkB' | 'pfkC')) & (('gapA.NAD+' | 'gapA.NADP+')) & (('pgk')) & (('gpmA' | 'gpmB' | 'gpmI' | 'apgM')) & (('eno')) & (('pyk'));(('scrK' | 'fruK'))
Resource Acquisition:Substrate degradation:simple compound degradation:carbohydrate degradation:ED pathway	(('glk')) & (('gpd')) & (('pgl_1' | 'pgl_2')) & (('edd')) & (('eda' | 'dgaF' | 'kdpgA'))
Resource Acquisition:Substrate degradation:simple compound degradation:carbohydrate degradation:EMP pathway	(('glk')) & (('pgi' | 'pgi_arch' | 'pgi-tal' | 'pgi-pmi')) & (('pfk' | 'pfkA' | 'pfkB' | 'pfkC')) & (('gapA.NAD+' | 'gapA.NADP+')) & (('pgk')) & (('gpmA' | 'gpmB' | 'gpmI' | 'apgM')) & (('eno')) & (('pyk'))`

Reply:
Why not parse out the list of genes here and just put that list in the file?

Reply:
you need the logic to match the result boolean string something like `(1 | (0 & 0) | 1)`.

Reply:
I was under the impression we'd just take all the genes for the trait, grab their score from the genes_detected table, and show them all. I assume there's some sort of gene score cut off to generate the boolean?

Reply:
That was my understanding. That's easier I presume (but LMK if I'm wrong), so let's assume we're doing it that way and I'll ping @psdehal

Reply:
The score cutoff is encoded in the hmm models (TC option). The cutoffs are benchmarked independently as explained in the paper). The hits table already has those cutoffs imposed, so anything that doesn't show up is a negative.

Reply:
Thanks @ukaraoz.

@Tianhao-Gu so IIUC, sounds like you can just grab all the genes from the trait rules, ignoring the boolean rules in the RDS files, and pull the gene scores from the genes detected table, dropping any non-present genes from the calculation script output

Reply:
The hmm hit list is saved in the rds output file as well.

Reply:
Looks ok to me although you might want to update the header

Reply:
👍

Reply:
:+1: